File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/Events.scala
package org.infinispan.server.hotrod

import org.infinispan.commons.util.Util
import org.infinispan.server.hotrod.OperationResponse._

/**
 * @author Galder Zamarreño
 */
object Events {

   abstract class Event(val version: Byte, val messageId: Long, val op: OperationResponse, val listenerId: Bytes, val isRetried: Boolean)

   case class KeyEvent(
           override val version: Byte,
           override val messageId: Long,
           override val listenerId: Bytes,
           override val isRetried: Boolean,
           key: Bytes)
         extends Event(version, messageId, CacheEntryRemovedEventResponse, listenerId, isRetried) {
      override def toString: String = {
         new StringBuilder().append("KeyEvent").append("{")
                 .append("version=").append(version)
                 .append(", messageId=").append(messageId)
                 .append(", listenerId=").append(Util.printArray(listenerId, false))
                 .append(", key=").append(Util.toStr(key))
                 .append(", isRetried=").append(isRetried)
                 .append("}").toString()
      }
   }

   case class KeyWithVersionEvent(
           override val version: Byte,
           override val messageId: Long,
           override val op: OperationResponse,
           override val listenerId: Bytes,
           override val isRetried: Boolean,
           key: Bytes, dataVersion: Long)
         extends Event(version, messageId, op, listenerId, isRetried) {
      override def toString: String = {
         new StringBuilder().append("KeyWithVersionEvent").append("{")
                 .append("version=").append(version)
                 .append(", messageId=").append(messageId)
                 .append(", op=").append(op)
                 .append(", listenerId=").append(Util.toStr(listenerId))
                 .append(", key=").append(Util.toStr(key))
                 .append(", dataVersion=").append(dataVersion)
                 .append(", isRetried=").append(isRetried)
                 .append("}").toString()
      }
   }

   case class CustomEvent(
               override val version: Byte,
               override val messageId: Long,
               override val op: OperationResponse,
               override val listenerId: Bytes,
               override val isRetried: Boolean,
               eventData: Bytes)
           extends Event(version, messageId, op, listenerId, isRetried) {
      override def toString: String = {
         new StringBuilder().append("CustomEvent").append("{")
                 .append("version=").append(version)
                 .append(", messageId=").append(messageId)
                 .append(", op=").append(op)
                 .append(", listenerId=").append(Util.toStr(listenerId, false))
                 .append(", event=").append(Util.toStr(eventData, false))
                 .append(", isRetried=").append(isRetried)
                 .append("}").toString()
      }
   }

}

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/Events.java
package org.infinispan.server.hotrod;

import java.util.concurrent.CompletableFuture;

import org.infinispan.commons.util.Util;
import org.infinispan.server.hotrod.transport.ExtendedByteBuf;

import io.netty.buffer.ByteBuf;

/**
 * @author Galder Zamarreño
 */
class Events {

   abstract static class Event {
      protected final byte version;
      protected final long messageId;
      protected final HotRodOperation op;
      protected final byte[] listenerId;
      protected final boolean isRetried;
      protected final byte marker;

      // Delays the operation that generated this event
      protected final CompletableFuture<Void> eventFuture;

      protected Event(byte version, long messageId, HotRodOperation op, byte[] listenerId, boolean isRetried, byte marker,
                      CompletableFuture<Void> eventFuture) {
         this.version = version;
         this.messageId = messageId;
         this.op = op;
         this.listenerId = listenerId;
         this.isRetried = isRetried;
         this.marker = marker;
         this.eventFuture = eventFuture;
      }

      abstract void writeEvent(ByteBuf buf);

      void defaultEvent(ByteBuf buf) {
         buf.writeByte(marker); // custom marker
         buf.writeByte(isRetried ? 1 : 0);
      }
   }

   static class KeyEvent extends Event {
      protected final byte[] key;

      protected KeyEvent(byte version, long messageId, HotRodOperation op, byte[] listenerId, boolean isRetried,
                         byte[] key, CompletableFuture<Void> eventFuture) {
         super(version, messageId, op, listenerId, isRetried, (byte) 0, eventFuture);
         this.key = key;
      }

      @Override
      public String toString() {
         return "KeyEvent{" +
               "version=" + version +
               ", messageId=" + messageId +
               ", op=" + op +
               ", listenerId=" + Util.printArray(listenerId, false) +
               ", isRetried=" + isRetried +
               ", key=" + Util.toStr(key) +
               '}';
      }

      @Override
      void writeEvent(ByteBuf buf) {
         defaultEvent(buf);
         ExtendedByteBuf.writeRangedBytes(key, buf);
      }
   }

   static class KeyWithVersionEvent extends Event {
      protected final byte[] key;
      protected final long dataVersion;

      protected KeyWithVersionEvent(byte version, long messageId, HotRodOperation op, byte[] listenerId, boolean isRetried,
                                    byte[] key, long dataVersion, CompletableFuture<Void> eventFuture) {
         super(version, messageId, op, listenerId, isRetried, (byte) 0, eventFuture);
         this.key = key;
         this.dataVersion = dataVersion;
      }

      @Override
      public String toString() {
         return "KeyWithVersionEvent{" +
               "version=" + version +
               ", messageId=" + messageId +
               ", op=" + op +
               ", listenerId=" + Util.printArray(listenerId, false) +
               ", isRetried=" + isRetried +
               ", key=" + Util.toStr(key) +
               ", dataVersion=" + dataVersion +
               '}';
      }

      @Override
      void writeEvent(ByteBuf buf) {
         defaultEvent(buf);
         ExtendedByteBuf.writeRangedBytes(key, buf);
         buf.writeLong(dataVersion);
      }
   }

   static class CustomEvent extends Event {
      protected final byte[] eventData;

      protected CustomEvent(byte version, long messageId, HotRodOperation op, byte[] listenerId, boolean isRetried,
                            byte[] eventData, CompletableFuture<Void> eventFuture) {
         super(version, messageId, op, listenerId, isRetried, (byte) 1, eventFuture);
         this.eventData = eventData;
      }

      @Override
      public String toString() {
         return "CustomEvent{" +
               "version=" + version +
               ", messageId=" + messageId +
               ", op=" + op +
               ", listenerId=" + Util.printArray(listenerId, false) +
               ", isRetried=" + isRetried +
               ", event=" + Util.toStr(eventData) +
               '}';
      }

      @Override
      void writeEvent(ByteBuf buf) {
         defaultEvent(buf);
         ExtendedByteBuf.writeRangedBytes(eventData, buf);
      }
   }

   static class CustomRawEvent extends Event {
      protected final byte[] eventData;

      protected CustomRawEvent(byte version, long messageId, HotRodOperation op, byte[] listenerId, boolean isRetried,
                               byte[] eventData, CompletableFuture<Void> eventFuture) {
         super(version, messageId, op, listenerId, isRetried, (byte) 2, eventFuture);
         this.eventData = eventData;
      }

      @Override
      public String toString() {
         return "CustomRawEvent{" +
               "version=" + version +
               ", messageId=" + messageId +
               ", op=" + op +
               ", listenerId=" + Util.printArray(listenerId, false) +
               ", isRetried=" + isRetried +
               ", event=" + Util.toStr(eventData) +
               '}';
      }

      @Override
      void writeEvent(ByteBuf buf) {
         defaultEvent(buf);
         ExtendedByteBuf.writeRangedBytes(eventData, buf);
      }
   }
}
